File: CS/GridViewFixedRows/GridFixedRowsPainter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using System.Data;
using DevExpress.XtraGrid.Views.Grid.Drawing;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Drawing;

namespace GVFixedRows
{
    class GridFixedRowsPainter : GridPainter
    {
        public GridFixedRowsPainter(GridView view) : base(view) { }

        protected override void DrawRows(GridViewDrawArgs e)
        {
            GridViewFixedRows gv = this.View as GridViewFixedRows;
            GridViewRectsFixedRows rectangles = e.ViewInfo.ViewRects as GridViewRectsFixedRows;
            int[] columnsWidths = new int[e.ViewInfo.ColumnsInfo.Count];

            for (int i = 0; i < e.ViewInfo.ColumnsInfo.Count; i++)
            {
                GridColumnInfoArgs ci = e.ViewInfo.ColumnsInfo[i];
                columnsWidths[i] = ci.Bounds.Width;
            }

            Rectangle r = new Rectangle();
            r.Height = e.ViewInfo.ColumnRowHeight;
            
            int rowCount = 0;
            foreach (DataRow row in gv.FixedRows)
            {
                r.X = rectangles.FixedRows.X - 1;
                for (int i = 0; i < columnsWidths.Length; i++)
                {
                    r.Y = rectangles.FixedRows.Y + e.ViewInfo.ColumnRowHeight * rowCount - 1;
                    r.Width = columnsWidths[i];

                    if (i == 0)
                        DrawIndicatorCell(e, r);
                    else
                    {
                        r.X += columnsWidths[i - 1];
                        DrawDataCell(e, gv, r, row[i-1].ToString());
                    }
                }
                rowCount++;
            }
            base.DrawRows(e);
        }

        void DrawDataCell(GridViewDrawArgs e, GridViewFixedRows gv, Rectangle r, string text)
        {
            e.Graphics.FillRectangle(e.Cache.GetSolidBrush(Color.LightGray), r);
            e.Graphics.DrawRectangle(e.Cache.GetPen(Color.Gray), r);
            
            Rectangle textR = r;
            textR.Inflate(-4, -2);

            StringFormat sf = gv.Appearance.Row.GetStringFormat();
            if (IsNumeric(text))
                sf.Alignment = StringAlignment.Far;
            else
                sf.Alignment = StringAlignment.Near;
            e.Graphics.DrawString(text, gv.Appearance.Row.Font, e.Cache.GetSolidBrush(Color.Black), textR, sf);
        }

        void DrawIndicatorCell(GridViewDrawArgs e, Rectangle r)
        {
            e.Graphics.FillRectangle(e.Cache.GetSolidBrush(Color.LightGray), r);
            e.Graphics.DrawRectangle(e.Cache.GetPen(Color.Gray), r);
        }

        bool IsNumeric(string str)
        {
            double num;
            return double.TryParse(str, out num);
        }
    }
}

File: VB/GridViewFixedRows/GridFixedRowsPainter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Drawing
Imports System.Data
Imports DevExpress.XtraGrid.Views.Grid.Drawing
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Drawing

Namespace GVFixedRows
	Friend Class GridFixedRowsPainter
		Inherits GridPainter
		Public Sub New(ByVal view As GridView)
			MyBase.New(view)
		End Sub

		Protected Overrides Sub DrawRows(ByVal e As GridViewDrawArgs)
			Dim gv As GridViewFixedRows = TryCast(Me.View, GridViewFixedRows)
			Dim rectangles As GridViewRectsFixedRows = TryCast(e.ViewInfo.ViewRects, GridViewRectsFixedRows)
			Dim columnsWidths(e.ViewInfo.ColumnsInfo.Count - 1) As Integer

			For i As Integer = 0 To e.ViewInfo.ColumnsInfo.Count - 1
				Dim ci As GridColumnInfoArgs = e.ViewInfo.ColumnsInfo(i)
				columnsWidths(i) = ci.Bounds.Width
			Next i

			Dim r As New Rectangle()
			r.Height = e.ViewInfo.ColumnRowHeight

			Dim rowCount As Integer = 0
			For Each row As DataRow In gv.FixedRows
				r.X = rectangles.FixedRows.X - 1
				For i As Integer = 0 To columnsWidths.Length - 1
					r.Y = rectangles.FixedRows.Y + e.ViewInfo.ColumnRowHeight * rowCount - 1
					r.Width = columnsWidths(i)

					If i = 0 Then
						DrawIndicatorCell(e, r)
					Else
						r.X += columnsWidths(i - 1)
						DrawDataCell(e, gv, r, row(i-1).ToString())
					End If
				Next i
				rowCount += 1
			Next row
			MyBase.DrawRows(e)
		End Sub

		Private Sub DrawDataCell(ByVal e As GridViewDrawArgs, ByVal gv As GridViewFixedRows, ByVal r As Rectangle, ByVal text As String)
			e.Graphics.FillRectangle(e.Cache.GetSolidBrush(Color.LightGray), r)
			e.Graphics.DrawRectangle(e.Cache.GetPen(Color.Gray), r)

			Dim textR As Rectangle = r
			textR.Inflate(-4, -2)

			Dim sf As StringFormat = gv.Appearance.Row.GetStringFormat()
			If IsNumeric(text) Then
				sf.Alignment = StringAlignment.Far
			Else
				sf.Alignment = StringAlignment.Near
			End If
			e.Graphics.DrawString(text, gv.Appearance.Row.Font, e.Cache.GetSolidBrush(Color.Black), textR, sf)
		End Sub

		Private Sub DrawIndicatorCell(ByVal e As GridViewDrawArgs, ByVal r As Rectangle)
			e.Graphics.FillRectangle(e.Cache.GetSolidBrush(Color.LightGray), r)
			e.Graphics.DrawRectangle(e.Cache.GetPen(Color.Gray), r)
		End Sub

		Private Function IsNumeric(ByVal str As String) As Boolean
			Dim num As Double
			Return Double.TryParse(str, num)
		End Function
	End Class
End Namespace
